src/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.cs

```diff
@@ -1034,29 +1044,17 @@ private void Grow(int requiredSize)
 
             int sizeHint = Math.Max(DefaultGrowthSize, requiredSize);
 
-            Debug.Assert(BytesPending != 0);
+            Debug.Assert(_currentIndex != 0);
 
             _memory = default;
 
             if (_stream != null)
             {
                 Debug.Assert(_arrayBufferWriter != null);
 
-                _arrayBufferWriter.Advance(BytesPending);
-                BytesPending = 0;
-
-#if BUILDING_INBOX_LIBRARY
-                _stream.Write(_arrayBufferWriter.WrittenSpan);
-#else
-                Debug.Assert(_arrayBufferWriter.WrittenMemory.Length == _arrayBufferWriter.WrittenCount);
-                bool result = MemoryMarshal.TryGetArray(_arrayBufferWriter.WrittenMemory, out ArraySegment<byte> underlyingBuffer);
-                Debug.Assert(result);
-                Debug.Assert(underlyingBuffer.Offset == 0);
-                Debug.Assert(_arrayBufferWriter.WrittenCount == underlyingBuffer.Count);
-                _stream.Write(underlyingBuffer.Array, underlyingBuffer.Offset, underlyingBuffer.Count);
-#endif
-                BytesCommitted += _arrayBufferWriter.WrittenCount;
-                _arrayBufferWriter.Clear();
+                _arrayBufferWriter.Advance(_currentIndex);
```

Comment:
Couldn't almost all of this change be avoided by just not calling Advance and just change the call to GetMemory to be BytesPending + sizeHint?

Reply:
I second Jeremy's question.

Reply:
I believe it can. That is the simpler approach I was initially trying to go for. Thanks, @bartonjs.

Reply:
**So, making such a change would mean that the user can only write up to 1 GB (without writing to the stream/flushing), even if the underlying buffer capacity is > 1 GB (say between 1 and 2 GB).**

Let's say you have written 1 GB so far and want to write a 150 MB JSON string token. If `BytesPending = 1 GB`, then `BytesPending + sizeHint of 150 MB > 1 GB`. And if we hit "Grow" in that case, the underlying buffer must be at least 1 GB (but less than `1 GB + sizeHint`, let's say its just above 1 GB). With what you propose, we end up asking for 1.15 GB (rather than 150 MB). The current growth scheme of **Max(double, what you asked for)** in ABW then throws OverflowException (since `1 GB + 1.15 GB > int.MaxValue`):
https://github.com/dotnet/corefx/blob/c707378f2a8ae1da5ebc5d02bbe2e0a3fd26a6a0/src/Common/src/System/Buffers/ArrayBufferWriter.cs#L177

In the case of IBW, and the change in this PR, that isn't the case (and the user can use a lot more of the available capacity). **Maybe supporting `> 1 GB` writes is an impractical edge case (where the user is doing way too much synchronous writing) and throwing is acceptable rather than complicating the logic.**

With the current change/PR, if you have written 1 GB and want to write a 150 MB string (and you hit grow), we ask for 150 MB, **not 1GB+** (which doubles the underlying buffer from 1 GB to 2 GB). That allows us to continue writing up to ~2 GB.

Reply:
Something like this would fail:
```C#
[Fact]
public void WriteLargeJsonToStreamWithoutFlushing()
{
    var largeArray = new char[150_000_000];
    largeArray.AsSpan().Fill('a');

    JsonEncodedText text1 = JsonEncodedText.Encode(largeArray.AsSpan(0, 7_500));
    JsonEncodedText text2 = JsonEncodedText.Encode(largeArray.AsSpan(0, 5_000));
    JsonEncodedText text3 = JsonEncodedText.Encode(largeArray.AsSpan(0, 150_000_000));

    using (var output = new MemoryStream())
    using (var writer = new Utf8JsonWriter(output))
    {
        writer.WriteStartArray();
        writer.WriteStringValue(text1);
        Assert.Equal(7503, writer.BytesPending);

        for (int i = 0; i < 30_000; i++)
        {
            writer.WriteStringValue(text2);
        }
        Assert.Equal(150097503, writer.BytesPending);

        for (int i = 0; i < 12; i++)
        {
            writer.WriteStringValue(text3); // throws OverFlow exception after the 6th write, when i  == 6
        }
        Assert.Equal(1950097539, writer.BytesPending);
    }
}

```text
System.Text.Json.Tests.Utf8JsonWriterTests.WriteLargeJsonToStreamWithoutFlushing [FAIL]
        System.OverflowException : Arithmetic operation resulted in an overflow.
        Stack Trace:
          E:\GitHub\Fork\corefx\src\Common\src\System\Buffers\ArrayBufferWriter.cs(177,0): at System.Buffers.ArrayBufferWriter`1.CheckAndResizeBuffer(Int32 sizeHint)
          E:\GitHub\Fork\corefx\src\Common\src\System\Buffers\ArrayBufferWriter.cs(130,0): at System.Buffers.ArrayBufferWriter`1.GetMemory(Int32 sizeHint)
          E:\GitHub\Fork\corefx\src\System.Text.Json\src\System\Text\Json\Writer\Utf8JsonWriter.cs(1046,0): at System.Text.Json.Utf8JsonWriter.Grow(Int32 requiredSize)
          E:\GitHub\Fork\corefx\src\System.Text.Json\src\System\Text\Json\Writer\Utf8JsonWriter.WriteValues.String.cs(267,0): at System.Text.Json.Utf8JsonWriter.WriteStringMinimized(ReadOnlySpan`1 escapedValue)
          E:\GitHub\Fork\corefx\src\System.Text.Json\src\System\Text\Json\Writer\Utf8JsonWriter.WriteValues.String.cs(253,0): at System.Text.Json.Utf8JsonWriter.WriteStringByOptions(ReadOnlySpan`1 utf8Value)
          E:\GitHub\Fork\corefx\src\System.Text.Json\src\System\Text\Json\Writer\Utf8JsonWriter.WriteValues.String.cs(26,0): at System.Text.Json.Utf8JsonWriter.WriteStringValueHelper(ReadOnlySpan`1 utf8Value)
          E:\GitHub\Fork\corefx\src\System.Text.Json\src\System\Text\Json\Writer\Utf8JsonWriter.WriteValues.String.cs(20,0): at System.Text.Json.Utf8JsonWriter.WriteStringValue(JsonEncodedText value)
          E:\GitHub\Fork\corefx\src\System.Text.Json\tests\Utf8JsonWriterTests.cs(184,0): at System.Text.Json.Tests.Utf8JsonWriterTests.WriteLargeJsonToStreamWithoutFlushing()
```
```